In [70]:
import pandas as pd
from sklearn.linear_model import LinearRegression
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [9]:
regular_season_outcomes = pd.read_csv('/content/drive/MyDrive/ChicagoBears/TeamStats/2025RegularSeason.csv')
regular_season_outcomes

,Rk,Gtm,Week,Date,Day,HomeAway,Opp,Rslt,Pts,PtsO,...,3DConv,3DAtt,4DConv,4DAtt,Pen,PenYds,FL,Int,TO,ToP
0,1.0,1.0,1.0,2025-09-08,Mon,NaN,MIN,L,24,27,...,3,12,0,1,12,127,1,0,1,32:53
1,2.0,2.0,2.0,2025-09-14,Sun,@,DET,L,21,52,...,7,13,0,2,8,50,1,1,2,29:40
2,3.0,3.0,3.0,2025-09-21,Sun,NaN,DAL,W,31,14,...,8,14,1,1,6,41,0,0,0,27:02
3,4.0,4.0,4.0,2025-09-28,Sun,@,LVR,W,25,24,...,7,16,1,1,8,60,0,1,1,33:10
4,5.0,5.0,6.0,2025-10-13,Mon,@,WAS,W,25,24,...,3,10,0,1,9,84,0,0,0,28:26
5,6.0,6.0,7.0,2025-10-19,Sun,NaN,NOR,W,26,14,...,3,12,1,1,10,92,0,1,1,36:19
6,7.0,7.0,8.0,2025-10-26,Sun,@,BAL,L,16,30,...,6,13,0,1,11,79,0,1,1,29:30
7,8.0,8.0,9.0,2025-11-02,Sun,@,CIN,W,47,42,...,8,15,2,2,7,43,0,0,0,34:30
8,9.0,9.0,10.0,2025-11-09,Sun,NaN,NYG,W,24,20,...,3,11,1,4,3,25,0,0,0,28:42
9,10.0,10.0,11.0,2025-11-16,Sun,@,MIN,W,19,17,...,7,18,0,0,6,40,0,0,0,36:59


In [10]:
regular_season_outcomes.columns

Index(['Rk', 'Gtm', 'Week', 'Date', 'Day', 'HomeAway', 'Opp', 'Rslt', 'Pts',
       'PtsO', 'OT', 'Cmp', 'PassAtt', 'Cmp%', 'PassYds', 'PassTD', 'Pass_Y/A',
       'AY/A', 'Rate', 'Sk', 'SackYds', 'RushAtt', 'RushYds', 'RushTD',
       'Rush_Y/A', 'Ply', 'Tot', 'Y/P', 'FGA', 'FGM', 'XPA', 'XPM', 'Pnt',
       'PuntYds', 'FD_Pass', 'FD_Rsh', 'FD_Pen', '1stD', '3DConv', '3DAtt',
       '4DConv', '4DAtt', 'Pen', 'PenYds', 'FL', 'Int', 'TO', 'ToP'],
      dtype='object')

In [12]:
# Create a bar plot to visualize the average Time of Possession for wins and losses
plt.figure(figsize=(8, 6))
sns.barplot(x='Rslt', y='ToP_seconds', data=average_top_by_result, palette={'L': 'salmon', 'W': 'lightgreen'})
plt.title('Average Time of Possession for Wins vs. Losses')
plt.xlabel('Result')
plt.ylabel('Average ToP (Seconds)')
plt.show()

print(f"The difference in average Time of Possession between wins and losses is approximately {difference_in_top_mins:.2f} minutes for the Chicago Bears.")

NameError: name 'average_top_by_result' is not defined

<Figure size 800x600 with 0 Axes>

In [14]:
opponent_regular_season_outcomes = pd.read_csv('/content/drive/MyDrive/ChicagoBears/TeamStats/Opponent2025RegularSeason.csv')
opponent_regular_season_outcomes['ToP']

all_reg_season_outcomes = regular_season_outcomes.join(opponent_regular_season_outcomes, lsuffix = 'Chi', rsuffix = 'Opp')
all_ToP = all_reg_season_outcomes[['RsltChi','ToPChi','ToPOpp']]
all_ToP = all_ToP.iloc[0:17]
all_ToP

,RsltChi,ToPChi,ToPOpp
0,L,32:53,27:07
1,L,29:40,30:20
2,W,27:02,32:58
3,W,33:10,26:50
4,W,28:26,31:34
5,W,36:19,23:41
6,L,29:30,30:30
7,W,34:30,25:30
8,W,28:42,31:18
9,W,36:59,23:01


In [15]:
# Helper function to convert MM:SS string to seconds
def convert_to_seconds(time_str):
    if pd.isna(time_str): # Handle NaN values if any
        return np.nan
    parts = str(time_str).split(':')
    if len(parts) == 2:
        minutes, seconds = int(parts[0]), int(parts[1])
        return minutes * 60 + seconds
    return np.nan # Return NaN for invalid formats

# Convert 'ToPChi' and 'ToPOpp' to seconds
all_ToP['ToPChi_seconds'] = all_ToP['ToPChi'].apply(convert_to_seconds)
all_ToP['ToPOpp_seconds'] = all_ToP['ToPOpp'].apply(convert_to_seconds)

# Calculate the difference using a lambda function
all_ToP['ToP_Diff'] = all_ToP.apply(lambda row: row['ToPChi_seconds'] - row['ToPOpp_seconds'], axis=1)

# Display the updated DataFrame
display(all_ToP)

,RsltChi,ToPChi,ToPOpp,ToPChi_seconds,ToPOpp_seconds,ToP_Diff
0,L,32:53,27:07,1973,1627,346
1,L,29:40,30:20,1780,1820,-40
2,W,27:02,32:58,1622,1978,-356
3,W,33:10,26:50,1990,1610,380
4,W,28:26,31:34,1706,1894,-188
5,W,36:19,23:41,2179,1421,758
6,L,29:30,30:30,1770,1830,-60
7,W,34:30,25:30,2070,1530,540
8,W,28:42,31:18,1722,1878,-156
9,W,36:59,23:01,2219,1381,838


In [16]:
unexpected_wins = all_ToP[(all_ToP['RsltChi'] == 'W') & (all_ToP['ToP_Diff'] < 0)]
unexpected_wins_count = len(unexpected_wins)
unexpected_wins_count

5

In [17]:
expected_wins = all_ToP[(all_ToP['RsltChi'] == 'W') & (all_ToP['ToP_Diff'] > 0)]
expected_wins_count = len(expected_wins)
expected_wins_count

6

In [18]:
expected_losses = all_ToP[(all_ToP['RsltChi'] == 'L') & (all_ToP['ToP_Diff'] < 0)]
expected_losses_count = len(expected_losses)
expected_losses_count

4

In [19]:
unexpected_losses = all_ToP[(all_ToP['RsltChi'] == 'L') & (all_ToP['ToP_Diff'] > 0)]
unexpected_losses_count = len(unexpected_losses)
unexpected_losses_count

2

Based on the analysis, there's no certain evidence that time of possession alone dictates a victory or a loss for the Bears. It's intriguing that the Bears do not necessarily need a larger time of possession than their opponent to secure a win. However, it was observed that only two losses occurred when the Bears had more time of possession, which generally aligns with the assumption that a lower time of possession often leads to a loss.

In [20]:
all_reg_season_outcomes['ToP_Diff'] = all_ToP['ToP_Diff']
display(all_reg_season_outcomes.head())

,RkChi,GtmChi,WeekChi,DateChi,DayChi,HomeAwayChi,OppChi,RsltChi,PtsChi,PtsOChi,...,3DAttOpp,4DConvOpp,4DAttOpp,PenOpp,PenYdsOpp,FLOpp,IntOpp,TOOpp,ToPOpp,ToP_Diff
0,1.0,1.0,1.0,2025-09-08,Mon,NaN,MIN,L,24,27,...,12,0,0,8,50,0,1,1,27:07,346.0
1,2.0,2.0,2.0,2025-09-14,Sun,@,DET,L,21,52,...,10,1,1,3,28,0,0,0,30:20,-40.0
2,3.0,3.0,3.0,2025-09-21,Sun,NaN,DAL,W,31,14,...,11,2,3,4,25,1,3,4,32:58,-356.0
3,4.0,4.0,4.0,2025-09-28,Sun,@,LVR,W,25,24,...,8,0,1,5,36,1,3,4,26:50,380.0
4,5.0,5.0,6.0,2025-10-13,Mon,@,WAS,W,25,24,...,8,1,1,5,40,2,1,3,31:34,-188.0


In [21]:
unexpected_win = all_reg_season_outcomes[
    (all_reg_season_outcomes['RsltChi'] == 'W') &
    (all_reg_season_outcomes['ToP_Diff'] < 0)
]

unexpected_win

,RkChi,GtmChi,WeekChi,DateChi,DayChi,HomeAwayChi,OppChi,RsltChi,PtsChi,PtsOChi,...,3DAttOpp,4DConvOpp,4DAttOpp,PenOpp,PenYdsOpp,FLOpp,IntOpp,TOOpp,ToPOpp,ToP_Diff
2,3.0,3.0,3.0,2025-09-21,Sun,NaN,DAL,W,31,14,...,11,2,3,4,25,1,3,4,32:58,-356.0
4,5.0,5.0,6.0,2025-10-13,Mon,@,WAS,W,25,24,...,8,1,1,5,40,2,1,3,31:34,-188.0
8,9.0,9.0,10.0,2025-11-09,Sun,NaN,NYG,W,24,20,...,14,1,4,10,69,1,0,1,31:18,-156.0
10,11.0,11.0,12.0,2025-11-23,Sun,NaN,PIT,W,31,28,...,14,1,3,6,41,1,1,2,33:09,-378.0
14,15.0,15.0,16.0,2025-12-20,Sat,NaN,GNB,W,22,16,...,12,1,3,4,40,1,0,1,38:57,-764.0


Based on these preliminary findings, it appears that Time of Possession alone doesn't definitively determine a win or loss for the Bears, especially given the 'unexpected' outcomes. It's now time to broaden our analysis and investigate other features or statistics that might have a stronger correlation with game results and help better determine a Bears' win.

In [22]:
regular_season_outcomes

,Rk,Gtm,Week,Date,Day,HomeAway,Opp,Rslt,Pts,PtsO,...,3DConv,3DAtt,4DConv,4DAtt,Pen,PenYds,FL,Int,TO,ToP
0,1.0,1.0,1.0,2025-09-08,Mon,NaN,MIN,L,24,27,...,3,12,0,1,12,127,1,0,1,32:53
1,2.0,2.0,2.0,2025-09-14,Sun,@,DET,L,21,52,...,7,13,0,2,8,50,1,1,2,29:40
2,3.0,3.0,3.0,2025-09-21,Sun,NaN,DAL,W,31,14,...,8,14,1,1,6,41,0,0,0,27:02
3,4.0,4.0,4.0,2025-09-28,Sun,@,LVR,W,25,24,...,7,16,1,1,8,60,0,1,1,33:10
4,5.0,5.0,6.0,2025-10-13,Mon,@,WAS,W,25,24,...,3,10,0,1,9,84,0,0,0,28:26
5,6.0,6.0,7.0,2025-10-19,Sun,NaN,NOR,W,26,14,...,3,12,1,1,10,92,0,1,1,36:19
6,7.0,7.0,8.0,2025-10-26,Sun,@,BAL,L,16,30,...,6,13,0,1,11,79,0,1,1,29:30
7,8.0,8.0,9.0,2025-11-02,Sun,@,CIN,W,47,42,...,8,15,2,2,7,43,0,0,0,34:30
8,9.0,9.0,10.0,2025-11-09,Sun,NaN,NYG,W,24,20,...,3,11,1,4,3,25,0,0,0,28:42
9,10.0,10.0,11.0,2025-11-16,Sun,@,MIN,W,19,17,...,7,18,0,0,6,40,0,0,0,36:59


In [23]:
pd.set_option('display.max_columns', None)

In [24]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [26]:
regular_season_outcomes

,Rk,Gtm,Week,Date,Day,HomeAway,Opp,Rslt,Pts,PtsO,OT,Cmp,PassAtt,Cmp%,PassYds,PassTD,Pass_Y/A,AY/A,Rate,Sk,SackYds,RushAtt,RushYds,RushTD,Rush_Y/A,Ply,Tot,Y/P,FGA,FGM,XPA,XPM,Pnt,PuntYds,FD_Pass,FD_Rsh,FD_Pen,1stD,3DConv,3DAtt,4DConv,4DAtt,Pen,PenYds,FL,Int,TO,ToP
0,1.0,1.0,1.0,2025-09-08,Mon,NaN,MIN,L,24,27,NaN,21,35,60.0,198,1,5.7,6.2,85.2,2,12,26,119,1,4.6,63,317,5.03,2,1,3,3,6,266,10,7,3,20,3,12,0,1,12,127,1,0,1,32:53
1,2.0,2.0,2.0,2025-09-14,Sun,@,DET,L,21,52,NaN,21,33,63.6,205,2,6.2,6.1,88.6,4,29,27,134,1,5.0,64,339,5.30,0,0,3,3,4,212,11,7,1,19,7,13,0,2,8,50,1,1,2,29:40
2,3.0,3.0,3.0,2025-09-21,Sun,NaN,DAL,W,31,14,NaN,19,28,67.9,298,4,10.6,13.5,142.6,0,0,29,87,0,3.0,57,385,6.75,1,1,4,4,4,176,11,7,1,19,8,14,1,1,6,41,0,0,0,27:02
3,4.0,4.0,4.0,2025-09-28,Sun,@,LVR,W,25,24,NaN,22,37,59.5,202,1,5.5,4.8,72.1,1,10,26,69,1,2.7,64,271,4.23,4,4,1,1,3,148,10,5,1,16,7,16,1,1,8,60,0,1,1,33:10
4,5.0,5.0,6.0,2025-10-13,Mon,@,WAS,W,25,24,NaN,17,29,58.6,236,1,8.1,8.8,96.3,3,16,27,150,1,5.6,59,386,6.54,5,4,1,1,2,106,10,8,1,19,3,10,0,1,9,84,0,0,0,28:26
5,6.0,6.0,7.0,2025-10-19,Sun,NaN,NOR,W,26,14,NaN,15,26,57.7,164,0,6.3,4.6,60.4,1,8,40,222,2,5.6,67,386,5.76,4,4,2,2,3,129,10,13,0,23,3,12,1,1,10,92,0,1,1,36:19
6,7.0,7.0,8.0,2025-10-26,Sun,@,BAL,L,16,30,NaN,25,38,65.8,276,0,7.3,6.1,76.2,1,9,23,96,1,4.2,62,372,6.00,4,3,1,1,2,108,13,5,1,19,6,13,0,1,11,79,0,1,1,29:30
7,8.0,8.0,9.0,2025-11-02,Sun,@,CIN,W,47,42,NaN,22,37,59.5,293,4,7.9,10.1,120.7,2,9,37,283,2,7.6,76,576,7.58,3,2,5,5,2,117,17,11,2,30,8,15,2,2,7,43,0,0,0,34:30
8,9.0,9.0,10.0,2025-11-09,Sun,NaN,NYG,W,24,20,NaN,20,36,55.6,220,1,6.1,6.7,83.1,0,0,28,171,2,6.1,64,391,6.11,1,1,3,3,3,135,11,8,4,23,3,11,1,4,3,25,0,0,0,28:42
9,10.0,10.0,11.0,2025-11-16,Sun,@,MIN,W,19,17,NaN,16,32,50.0,180,0,5.6,5.6,67.2,2,13,39,140,1,3.6,73,320,4.38,5,4,1,1,6,286,6,13,0,19,7,18,0,0,6,40,0,0,0,36:59


In [27]:
week1chi_drives = pd.read_csv('/content/drive/MyDrive/ChicagoBears/Drives/week1chi.csv')
week1chi_drives

,#,Quarter,Time,LOS,Plays,Length,Net Yds,Result
0,1,1,12:45,CHI 39,10,5:41,61,Touchdown
1,2,1,4:51,CHI 25,5,3:53,17,Punt
2,3,2,12:48,CHI 39,5,3:53,37,Downs
3,4,2,6:52,CHI 35,3,1:00,2,Punt
4,5,2,4:29,CHI 41,9,4:05,35,Field Goal
5,7,3,15:00,CHI 22,3,0:32,0,Punt
6,8,3,10:08,CHI 20,5,2:54,23,Punt
7,9,3,5:46,CHI 24,11,5:51,44,Missed FG
8,10,4,12:13,CHI 25,3,1:36,-9,Punt
9,11,4,9:46,CHI 35,5,2:22,8,Punt


In [39]:
total_plays = week1chi_drives['Plays'].sum()
total_yards = week1chi_drives['Net Yds'].sum()
average_yards_per_play = total_yards / total_plays
average_yards_per_play.round(2).astype(str) + ' yards per play -- Chicago'


'4.57 yards per play -- Chicago'

In [28]:
week1min_drives = pd.read_csv('/content/drive/MyDrive/ChicagoBears/Drives/week1min.csv')
week1min_drives

,#,Quarter,Time,LOS,Plays,Length,Net Yds,Result
0,1,1,15:00,MIN 25,3,2:15,7,Punt
1,2,1,7:04,MIN 24,3,2:13,4,Punt
2,3,1,0:58,MIN 19,7,3:10,67,Field Goal
3,4,2,8:55,MIN 25,3,2:03,9,Punt
4,5,2,5:52,MIN 24,3,1:23,-8,Punt
5,6,2,0:24,MIN 31,2,0:18,28,Field Goal
6,7,3,14:28,CHI 49,3,1:37,17,Interception
7,8,3,12:51,MIN 28,3,2:43,-10,Punt
8,9,3,7:14,MIN 23,3,1:28,8,Punt
9,10,4,14:55,MIN 40,6,2:42,60,Touchdown


In [40]:
total_plays = week1min_drives['Plays'].sum()
total_yards = week1min_drives['Net Yds'].sum()
average_yards_per_play = total_yards / total_plays
average_yards_per_play.round(2).astype(str) + ' yards per play -- Minnesota'


'5.98 yards per play -- Minnesota'

In [46]:
week2chi_drives = pd.read_csv('/content/drive/MyDrive/ChicagoBears/Drives/week2/chi.csv')
week2opp_drives = pd.read_csv('/content/drive/MyDrive/ChicagoBears/Drives/week2/opp.csv')

In [48]:
total_plays2 = week2chi_drives['Plays'].sum()
total_yards2 = week2chi_drives['Net Yds'].sum()
average_yards_per_play2 = total_yards2 / total_plays2
average_yards_per_play2.round(2).astype(str)

np.str_('4.73')

In [49]:
total_plays2o = week2opp_drives['Plays'].sum()
total_yards2o = week2opp_drives['Net Yds'].sum()
yards_per_play = total_yards2o / total_plays2o
yards_per_play.round(2).astype(str)

np.str_('8.52')

In [50]:
yards_per_play = pd.DataFrame(columns=['Chi','Opp','Win'])
yards_per_play.loc[0] = ['4.57', '5.98',False]
yards_per_play.loc[1] = ['4.73','8.52',False]


In [52]:
##Used claude and gemini to scrape the rest of the 32 csv tables from pro-football-reference.com and import all csv files and run the same code I ran for the first three weeks.

import os

# Path to the base directory for drives
base_drives_path = '/content/drive/MyDrive/ChicagoBears/Drives/'

# Function to calculate yards per play from a drive DataFrame
def get_yards_per_play(df_drives):
    total_plays = df_drives['Plays'].sum()
    total_yards = df_drives['Net Yds'].sum()
    if total_plays == 0:
        return 0.0
    return total_yards / total_plays

# Initialize a new DataFrame to store the results for all weeks
all_weeks_yards_per_play = pd.DataFrame(columns=['Week', 'Chi', 'Opp', 'Win'])

# Iterate from Week 1 to Week 17 (assuming 17 regular season games)
for week_num in range(1, 18):
    chi_ypp = None
    opp_ypp = None
    win_status = None

    # Get win/loss status for the current week from regular_season_outcomes
    week_outcome = regular_season_outcomes[regular_season_outcomes['Week'] == float(week_num)]
    if not week_outcome.empty:
        win_status = (week_outcome['Rslt'].iloc[0] == 'W')

    # Construct file paths based on observed patterns
    # Pattern 1: weekX/chi.csv and weekX/opp.csv
    week_dir_path = os.path.join(base_drives_path, f'week{week_num}')
    chi_file_path_pattern1 = os.path.join(week_dir_path, 'chi.csv')
    opp_file_path_pattern1 = os.path.join(week_dir_path, 'opp.csv')

    # Pattern 2: weekXchi.csv and weekXmin.csv (for week 1)
    chi_file_path_pattern2 = os.path.join(base_drives_path, f'week{week_num}chi.csv')
    opp_file_path_pattern2 = os.path.join(base_drives_path, f'week{week_num}min.csv') # Assuming 'min' for week 1 opponent

    # Determine which pattern applies and read the files
    if os.path.exists(chi_file_path_pattern1) and os.path.exists(opp_file_path_pattern1):
        chi_df = pd.read_csv(chi_file_path_pattern1)
        opp_df = pd.read_csv(opp_file_path_pattern1)
    elif os.path.exists(chi_file_path_pattern2) and os.path.exists(opp_file_path_pattern2):
        chi_df = pd.read_csv(chi_file_path_pattern2)
        opp_df = pd.read_csv(opp_file_path_pattern2)
    else:
        # Skip if files for the week are not found in either pattern
        continue

    # Calculate Yards Per Play
    chi_ypp = round(get_yards_per_play(chi_df), 2)
    opp_ypp = round(get_yards_per_play(opp_df), 2)

    # Add the data to the new DataFrame
    if chi_ypp is not None and opp_ypp is not None and win_status is not None:
        all_weeks_yards_per_play.loc[len(all_weeks_yards_per_play)] = [week_num, str(chi_ypp), str(opp_ypp), win_status]

display(all_weeks_yards_per_play)

,Week,Chi,Opp,Win
0,1,4.57,5.98,False
1,2,4.73,8.52,False
2,3,6.47,5.68,True
3,4,2.99,6.54,True
4,6,5.73,5.3,True
5,7,4.94,5.46,True
6,8,5.17,5.53,False
7,9,7.2,6.87,True
8,10,6.46,5.78,True
9,11,3.65,4.82,True


In [57]:
all_weeks_yards_per_play['YPP_Diff'] = all_weeks_yards_per_play.apply(
    lambda row: float(row['Chi']) - float(row['Opp']), axis=1
)

all_weeks_yards_per_play.sort_values(by='Win', ascending=False)

,Week,Chi,Opp,Win,YPP_Diff
2,3,6.47,5.68,True,0.79
3,4,2.99,6.54,True,-3.55
4,6,5.73,5.3,True,0.43
5,7,4.94,5.46,True,-0.52
7,9,7.2,6.87,True,0.33
8,10,6.46,5.78,True,0.68
9,11,3.65,4.82,True,-1.17
10,12,4.9,5.12,True,-0.22
11,13,4.81,5.6,True,-0.79
13,15,5.18,3.18,True,2.00


In [59]:
# Regular-season scores (Bears, Opp). Week 5 was the bye, so it's skipped.
scores = {
    1:  (24, 27),
    2:  (21, 52),
    3:  (31, 14),
    4:  (25, 24),
    6:  (25, 24),
    7:  (26, 14),
    8:  (16, 30),
    9:  (47, 42),
    10: (24, 20),
    11: (19, 17),
    12: (31, 28),
    13: (24, 15),
    14: (21, 28),
    15: (31,  3),
    16: (22, 16),
    17: (38, 42),
    18: (16, 19),
}

# Initialize the columns first (so .loc has somewhere to write)
all_weeks_yards_per_play['chi_pts'] = pd.NA
all_weeks_yards_per_play['opp_pts'] = pd.NA

# Use .loc with a boolean mask on the Week column — index order doesn't matter
for week, (chi, opp) in scores.items():
    mask = all_weeks_yards_per_play['Week'] == week
    all_weeks_yards_per_play.loc[mask, 'chi_pts'] = chi
    all_weeks_yards_per_play.loc[mask, 'opp_pts'] = opp

all_weeks_yards_per_play

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts
0,1,4.57,5.98,False,-1.41,24,27
1,2,4.73,8.52,False,-3.79,21,52
2,3,6.47,5.68,True,0.79,31,14
3,4,2.99,6.54,True,-3.55,25,24
4,6,5.73,5.3,True,0.43,25,24
5,7,4.94,5.46,True,-0.52,26,14
6,8,5.17,5.53,False,-0.36,16,30
7,9,7.2,6.87,True,0.33,47,42
8,10,6.46,5.78,True,0.68,24,20
9,11,3.65,4.82,True,-1.17,19,17


In [60]:
all_weeks_yards_per_play['point_diff'] = all_weeks_yards_per_play.apply(
    lambda row: row['chi_pts'] - row['opp_pts'], axis=1
)

In [64]:
all_weeks_yards_per_play

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff
0,1,4.57,5.98,False,-1.41,24,27,-3
1,2,4.73,8.52,False,-3.79,21,52,-31
2,3,6.47,5.68,True,0.79,31,14,17
3,4,2.99,6.54,True,-3.55,25,24,1
4,6,5.73,5.3,True,0.43,25,24,1
5,7,4.94,5.46,True,-0.52,26,14,12
6,8,5.17,5.53,False,-0.36,16,30,-14
7,9,7.2,6.87,True,0.33,47,42,5
8,10,6.46,5.78,True,0.68,24,20,4
9,11,3.65,4.82,True,-1.17,19,17,2


In [69]:
all_weeks_yards_per_play[['YPP_Diff', 'point_diff']].corr()

,YPP_Diff,point_diff
YPP_Diff,1.000000,0.714966
point_diff,0.714966,1.000000


For the Chicago Bears, every loss occurred when they had lower yards per play than their opponent, showing that efficiency disadvantage → loss, regardless of total yardage or possession.

### 1. Absolute YPP doesn’t guarantee a win
* The Bears don’t need a specific YPP threshold to win
* High YPP games can still be losses (if opponent is better)

### 2. Relative YPP is highly predictive
* When: YPP_diff = Bears YPP - Opponent YPP < 0

→ Loss

“Yards-per-play differential showed a strong positive correlation (r = 0.715) with point differential, indicating that relative efficiency is a key driver not only of game outcomes but also of margin of victory.”

In [71]:
##Lets create a simple regression model to further analyze our data

X = all_weeks_yards_per_play[['YPP_Diff']]
y = all_weeks_yards_per_play['point_diff']

In [72]:
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [79]:
all_weeks_yards_per_play["predicted_point_diff"] = model.predict(X)

In [89]:
print("Coefficient:", model.coef_[0])
print("Intercept:", model.intercept_)

score = model.score(X, y)
print("R^2:", score)

Coefficient: 6.249967458739539
Intercept: 5.4609185040392045
R^2: 0.5111757056958139


In [80]:
all_weeks_yards_per_play["error"] = all_weeks_yards_per_play["point_diff"] - all_weeks_yards_per_play["predicted_point_diff"]

In [81]:
all_weeks_yards_per_play

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff,predicted_point_diff,error
0,1,4.57,5.98,False,-1.41,24,27,-3,-3.351536,0.351536
1,2,4.73,8.52,False,-3.79,21,52,-31,-18.226458,-12.773542
2,3,6.47,5.68,True,0.79,31,14,17,10.398393,6.601607
3,4,2.99,6.54,True,-3.55,25,24,1,-16.726466,17.726466
4,6,5.73,5.3,True,0.43,25,24,1,8.148405,-7.148405
5,7,4.94,5.46,True,-0.52,26,14,12,2.210935,9.789065
6,8,5.17,5.53,False,-0.36,16,30,-14,3.210930,-17.210930
7,9,7.2,6.87,True,0.33,47,42,5,7.523408,-2.523408
8,10,6.46,5.78,True,0.68,24,20,4,9.710896,-5.710896
9,11,3.65,4.82,True,-1.17,19,17,2,-1.851543,3.851543


Yards-per-play differential explains general game outcomes, but fails in edge cases where situational factors like turnovers or red zone efficiency dominate.

In [82]:
mae = np.mean(np.abs(all_weeks_yards_per_play["error"]))

print("MAE:", mae)

MAE: 7.125000325412605


The model can currently expect to be off about 7 points. We should add another feature of importance to test and see if another feature can reduce mean error. Lower mean absoute error means our model is more accurate.

In [84]:
all_weeks_yards_per_play['chi_to'] = np.nan
all_weeks_yards_per_play['opp_to'] = np.nan
all_weeks_yards_per_play

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff,predicted_point_diff,error,chi_to,opp_to
0,1,4.57,5.98,False,-1.41,24,27,-3,-3.351536,0.351536,NaN,NaN
1,2,4.73,8.52,False,-3.79,21,52,-31,-18.226458,-12.773542,NaN,NaN
2,3,6.47,5.68,True,0.79,31,14,17,10.398393,6.601607,NaN,NaN
3,4,2.99,6.54,True,-3.55,25,24,1,-16.726466,17.726466,NaN,NaN
4,6,5.73,5.3,True,0.43,25,24,1,8.148405,-7.148405,NaN,NaN
5,7,4.94,5.46,True,-0.52,26,14,12,2.210935,9.789065,NaN,NaN
6,8,5.17,5.53,False,-0.36,16,30,-14,3.210930,-17.210930,NaN,NaN
7,9,7.2,6.87,True,0.33,47,42,5,7.523408,-2.523408,NaN,NaN
8,10,6.46,5.78,True,0.68,24,20,4,9.710896,-5.710896,NaN,NaN
9,11,3.65,4.82,True,-1.17,19,17,2,-1.851543,3.851543,NaN,NaN


In [88]:
# Loop through each week present in all_weeks_yards_per_play
for week_num in all_weeks_yards_per_play['Week'].unique():
    # Find the turnover data for the Bears for the current week
    # Note: 'Week' in regular_season_outcomes is float, so convert week_num to float for comparison
    chi_turnovers = regular_season_outcomes[regular_season_outcomes['Week'] == float(week_num)]['TO'].iloc[0]

    # Find the turnover data for the opponent for the current week
    opp_turnovers = opponent_regular_season_outcomes[opponent_regular_season_outcomes['Week'] == float(week_num)]['TO'].iloc[0]

    # Update the all_weeks_yards_per_play DataFrame
    mask = all_weeks_yards_per_play['Week'] == week_num
    all_weeks_yards_per_play.loc[mask, 'chi_to'] = chi_turnovers
    all_weeks_yards_per_play.loc[mask, 'opp_to'] = opp_turnovers

display(all_weeks_yards_per_play)

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff,predicted_point_diff,error,chi_to,opp_to,margin_to
0,1,4.57,5.98,False,-1.41,24,27,-3,-3.351536,0.351536,1.0,1.0,0.0
1,2,4.73,8.52,False,-3.79,21,52,-31,-18.226458,-12.773542,2.0,0.0,2.0
2,3,6.47,5.68,True,0.79,31,14,17,10.398393,6.601607,0.0,4.0,-4.0
3,4,2.99,6.54,True,-3.55,25,24,1,-16.726466,17.726466,1.0,4.0,-3.0
4,6,5.73,5.3,True,0.43,25,24,1,8.148405,-7.148405,0.0,3.0,-3.0
5,7,4.94,5.46,True,-0.52,26,14,12,2.210935,9.789065,1.0,4.0,-3.0
6,8,5.17,5.53,False,-0.36,16,30,-14,3.210930,-17.210930,1.0,0.0,1.0
7,9,7.2,6.87,True,0.33,47,42,5,7.523408,-2.523408,0.0,3.0,-3.0
8,10,6.46,5.78,True,0.68,24,20,4,9.710896,-5.710896,0.0,1.0,-1.0
9,11,3.65,4.82,True,-1.17,19,17,2,-1.851543,3.851543,0.0,2.0,-2.0


In [87]:
all_weeks_yards_per_play['margin_to'] = all_weeks_yards_per_play.apply(
    lambda row: row['chi_to'] - row['opp_to'], axis=1
)
all_weeks_yards_per_play

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff,predicted_point_diff,error,chi_to,opp_to,margin_to
0,1,4.57,5.98,False,-1.41,24,27,-3,-3.351536,0.351536,1.0,1.0,0.0
1,2,4.73,8.52,False,-3.79,21,52,-31,-18.226458,-12.773542,2.0,0.0,2.0
2,3,6.47,5.68,True,0.79,31,14,17,10.398393,6.601607,0.0,4.0,-4.0
3,4,2.99,6.54,True,-3.55,25,24,1,-16.726466,17.726466,1.0,4.0,-3.0
4,6,5.73,5.3,True,0.43,25,24,1,8.148405,-7.148405,0.0,3.0,-3.0
5,7,4.94,5.46,True,-0.52,26,14,12,2.210935,9.789065,1.0,4.0,-3.0
6,8,5.17,5.53,False,-0.36,16,30,-14,3.210930,-17.210930,1.0,0.0,1.0
7,9,7.2,6.87,True,0.33,47,42,5,7.523408,-2.523408,0.0,3.0,-3.0
8,10,6.46,5.78,True,0.68,24,20,4,9.710896,-5.710896,0.0,1.0,-1.0
9,11,3.65,4.82,True,-1.17,19,17,2,-1.851543,3.851543,0.0,2.0,-2.0


Given the insights from Yards Per Play, let's now incorporate 'turnover margin' as another crucial feature. Turnovers often significantly swing the momentum and outcome of a game, even when other statistics might suggest a different result. By including this, we aim to further enhance our model's predictive power and reduce the Mean Absolute Error (MAE), moving towards a more accurate understanding of what drives a Bears' win.

In [91]:
X = all_weeks_yards_per_play[['YPP_Diff','margin_to']]
y = all_weeks_yards_per_play['point_diff']

In [92]:
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [93]:
print("Coefficient:", model.coef_[0])
print("Intercept:", model.intercept_)

score = model.score(X, y)
print("R^2:", score)

Coefficient: 3.9486500294565587
Intercept: -2.0400223053856683
R^2: 0.783471169062565


In [97]:
all_weeks_yards_per_play["new_predict"] = model.predict(X)

In [101]:
all_weeks_yards_per_play["new_error"] = all_weeks_yards_per_play["point_diff"] - all_weeks_yards_per_play["new_predict"]

In [103]:
old_mae = np.mean(np.abs(all_weeks_yards_per_play["error"]))

new_mae = np.mean(np.abs(all_weeks_yards_per_play["new_error"]))

print("Old MAE:", old_mae)

print("New MAE:", new_mae)

Old MAE: 7.125000325412605
New MAE: 4.919604942028946


In [104]:
print("New R^2:", model.score(X, y))

New R^2: 0.783471169062565


While yards-per-play differential served as a strong baseline predictor, incorporating turnover margin significantly improved model performance, reducing prediction error by approximately 30%. This suggests that turnovers play a critical role in determining game outcomes beyond efficiency metrics alone.

In [105]:
all_weeks_yards_per_play.sort_values(by="new_error", ascending=False)

,Week,Chi,Opp,Win,YPP_Diff,chi_pts,opp_pts,point_diff,predicted_point_diff,error,chi_to,opp_to,margin_to,new_predict,new_error
11,13,4.81,5.6,True,-0.79,24,15,9,-0.681240,8.476556,1.0,2.0,-1.0,-0.681240,9.681240
13,15,5.18,3.18,True,2.00,31,3,28,19.291925,10.039147,0.0,3.0,-3.0,19.291925,8.708075
10,12,4.9,5.12,True,-0.22,31,28,3,-2.908725,-1.085926,2.0,2.0,0.0,-2.908725,5.908725
0,1,4.57,5.98,False,-1.41,24,27,-3,-7.607619,0.351536,1.0,1.0,0.0,-7.607619,4.607619
3,4,2.99,6.54,True,-3.55,25,24,1,-2.623082,17.726466,1.0,4.0,-3.0,-2.623082,3.623082
14,16,5.83,5.77,True,0.06,22,16,6,2.675113,0.164083,0.0,1.0,-1.0,2.675113,3.324887
5,7,4.94,5.46,True,-0.52,26,14,12,9.341327,9.789065,1.0,4.0,-3.0,9.341327,2.658673
12,14,4.57,6.04,False,-1.47,21,28,-7,-7.844538,-3.273466,1.0,1.0,0.0,-7.844538,0.844538
9,11,3.65,4.82,True,-1.17,19,17,2,2.296489,3.851543,0.0,2.0,-2.0,2.296489,-0.296489
8,10,6.46,5.78,True,0.68,24,20,4,5.123276,-5.710896,0.0,1.0,-1.0,5.123276,-1.123276


In [106]:
print(model.coef_)

[ 3.94865003 -4.47821583]


Having improved our model's accuracy by incorporating turnover margin, we now want to explore another potential feature: 'explosive plays'. To do this, we'll need to find and import an external play-by-play data source. We'll specifically look for plays with 20 or more yards gained, as these 'explosive plays' can significantly impact game momentum and outcomes, and we'll evaluate if this could be a valuable addition to our predictive model.

In [ ]:
url = "https://github.com/nflverse/nflverse-data/releases/download/pbp/play_by_play_2023.csv.gz"

df = pd.read_csv(url)

explosive = df[df["yards_gained"] >= 20]